tests/ballerina-spec-conformance-tests/src/test/java/org/ballerinalang/test/ReportGenerator.java

```diff
@@ -0,0 +1,194 @@
+    private static final Path TEST_DIR = BALLERINA_LANG_DIR.resolve("tests")
+                                                           .resolve("ballerina-spec-conformance-tests");
+    public static final Path REPORT_DIR = TEST_DIR.resolve("build").resolve("reports");
+    private static final String HTML_EXTENSION = ".html";
+    public static final String ERROR_KIND_TESTS_REPORT = "error_kind_tests_report_template.html";
+    public static final String SKIPPED_TESTS_REPORT = "skipped_tests_report_template.html";
+    public static final String FAILED_TESTS_REPORT = "failed_tests_report_template.html";
+    private static final String START_TABLE_ROW  = "<tr class=\"active-row\">";
+    private static final String END_TABLE_ROW  = "</tr>";
+    private List<Map<String, String>> detailsOfFailedTests;
+    private List<Map<String, String>> detailsOfSkippedTests;
+    private Map<String, List<Map<String, String>>> detailsOfErrorKindTests;
+
+    public ReportGenerator() {
+        this.detailsOfFailedTests = new ArrayList<>();
+        this.detailsOfSkippedTests = new ArrayList<>();
+        this.detailsOfErrorKindTests = new LinkedHashMap<>();
+    }
+
+    public void addDetailsOfSkippedTests(Map<String, String> detailsOfTests) {
+        detailsOfSkippedTests.add(detailsOfTests);
+    }
+
+    public void addDetailsOfFailedTests(Map<String, String> detailsOfTests) {
+        detailsOfFailedTests.add(detailsOfTests);
+    }
+
+    public void addDetailsOfErrorKindTests(Map<String, String> detailsOfTest) {
+        String fileName = detailsOfTest.get(FILENAME);
+        if (detailsOfErrorKindTests.containsKey(fileName)) {
+            detailsOfErrorKindTests.get(fileName).add(detailsOfTest);
+        } else {
+            List<Map<String, String>> details = new ArrayList<>();
+            details.add(detailsOfTest);
+            detailsOfErrorKindTests.put(fileName, details);
+        }
+    }
+
+    public void generateReport() throws IOException {
+        generateFailedTestsReports();
+        generateErrorVerificationTestReports();
+        generateSkippedTestReports();
+    }
+
+    private void generateFailedTestsReports() throws IOException {
+        if (detailsOfFailedTests.isEmpty()) {
+            return;
+        }
+        StringBuilder detailsOfTests = new StringBuilder();
+        for (Map<String, String> test : detailsOfFailedTests) {
+            String diagnostics = test.get(TestRunnerUtils.FORMAT_ERRORS);
+            if (diagnostics != null) {
+                detailsOfTests.append(generateFailedTestsDetails(test.get(FILENAME), test.get(KIND),
+                                      test.get(ABS_LINE_NUM), null, diagnostics, null));
+            } else {
+                detailsOfTests.append(generateFailedTestsDetails(test.get(FILENAME), test.get(KIND),
+                                      test.get(ACTUAL_LINE_NUM), test.get(EXPECTED_LINE_NUM), test.get(ACTUAL_VALUE),
```

Comment:
IMO, the more natural flow is to have the expected value first.

Reply:
you mean to change the expected and actual columns in the report.

Reply:
Yes.